clippy_lints/src/byte_char_slice.rs

```diff
@@ -0,0 +1,86 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use rustc_ast::ast::{BorrowKind, Expr, ExprKind, Mutability};
+use rustc_ast::token::{Lit, LitKind};
+use rustc_errors::Applicability;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for hard to read slices of byte characters, that could be more easily expressed as a
+    /// byte string.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Potentially makes the string harder to read.
+    ///
+    /// ### Example
+    /// ```rust
+    /// &[b'H', b'e', b'l', b'l', b'o']
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// b"Hello"
+    /// ```
+    #[clippy::version = "1.68.0"]
+    pub BYTE_CHAR_SLICE,
+    style,
+    "hard to read byte char slice"
+}
+declare_lint_pass!(ByteCharSlice => [BYTE_CHAR_SLICE]);
+
+impl EarlyLintPass for ByteCharSlice {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &Expr) {
+        if let Some(slice) = is_byte_char_slice(expr) && !expr.span.from_expansion() {
+            span_lint_and_sugg(
+                cx,
+                BYTE_CHAR_SLICE,
+                expr.span,
+                "can be more succinctly written as a byte str",
+                "try",
+                format!("b\"{slice}\""),
+                Applicability::MaybeIncorrect,
+            );
+        }
+    }
+}
+
+fn is_byte_char_slice(expr: &Expr) -> Option<String> {
+    if let ExprKind::AddrOf(BorrowKind::Ref, Mutability::Not, expr) = &expr.kind {
+        match &expr.kind {
+            ExprKind::Array(members) => {
+                let all_member_bytes = members.iter().all(|member| {
+                    matches!(
+                        &member.kind,
+                        ExprKind::Lit(Lit {
+                            kind: LitKind::Byte,
+                            ..
+                        })
+                    )
+                });
+
+                if !all_member_bytes || members.is_empty() {
+                    return None;
+                }
+
+                Some(
+                    members
+                        .iter()
+                        .map(|member| match &member.kind {
+                            ExprKind::Lit(Lit {
+                                kind: LitKind::Byte,
+                                symbol,
+                                ..
+                            }) => symbol.as_str(),
+                            _ => unreachable!(),
+                        })
+                        .collect::<Vec<_>>()
+                        .join(""),
+                )
```

Comment:
We can collect into an `Option<String>` from an iterator of `Option<&str>`s, that'll let us skip doing the validation in `all_member_bytes` separately

```suggestion
                members
                    .iter()
                    .map(|member| match &member.kind {
                        ExprKind::Lit(Lit {
                            kind: LitKind::Byte,
                            symbol,
                            ..
                        }) => Some(symbol.as_str()),
                        _ => None,
                    })
                    .collect()
```

Reply:
Done!